android/jni/com/mapswithme/maps/Framework.cpp

```diff
@@ -891,6 +892,51 @@ Java_com_mapswithme_maps_Framework_nativeGetRouteFollowingInfo(JNIEnv * env, jcl
   return result;
 }
 
+JNIEXPORT jintArray JNICALL
+Java_com_mapswithme_maps_Framework_nativeGenerateRouteAltitudeChartBits(JNIEnv * env, jclass, jint width, jint height)
+{
+  ::Framework * fr = frm();
+  ASSERT(fr, ());
+  vector<uint8_t> imageRGBAData;
+
+  if (!fr->GenerateRouteAltitudeChart(width, height, imageRGBAData))
+  {
+    LOG(LWARNING, ("Cann't generate route altitude image."));
+    return nullptr;
+  }
+
+  size_t const imageRGBADataSize = imageRGBAData.size();
+  if (imageRGBADataSize == 0)
+  {
+    LOG(LWARNING, ("GenerateRouteAltitudeChart returns true but the vector with altitude image bits is empty."));
+    return nullptr;
+  }
+
+  size_t const pxlCount = width * height;
+  if (maps::kAlitudeChartBPP * pxlCount != imageRGBAData.size())
+  {
+    LOG(LWARNING, ("Wrong size of vector with altitude image bits. Expected size:", pxlCount, ". Real size:", imageRGBAData.size()));
+    return nullptr;
+  }
+
+  jintArray imageRGBADataArray = env->NewIntArray(imageRGBADataSize);
+  ASSERT(imageRGBADataArray, ());
+  jint * arrayElements = env->GetIntArrayElements(imageRGBADataArray, 0);
+  ASSERT(arrayElements, ());
+
+  for (size_t i = 0; i < imageRGBADataSize; ++i)
+  {
+    size_t const shiftInBytes = i * maps::kAlitudeChartBPP;
+    arrayElements[i] = (imageRGBAData[shiftInBytes + 3] << 24) /* alpha */
+        | (imageRGBAData[shiftInBytes] << 16) /* red */
```

Comment:
There's another problem in this code. On the one hand jint is typedef on int (signed int). We have an array of them. On the other hand uint8_t is promoted to unsigned int before the shift and on the right side we have a big unsigned int. Then we assign unsigned int to int. According to the standard:

4.7 Integral conversions 
....
 If the destination type is signed, the value is unchanged if it can be represented in the destination type (and bit-field width); otherwise, the value is implementation-defined.